text/0000-app-image-extensions.md

```diff
@@ -0,0 +1,242 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+- *root buildpack* - a special case of buildpack that is run with privileges
+
+Application developers can use root buildpacks to extend their build and/or run images.
+
+We introduce a boolean `privileged` key in the `[buildpack]` table of `buildpack.toml`, which is defined as follows:
+
+```
+[buildpack]
+privileged = <boolean (default=false)>
+```
+
+When `privileged` is set to `true`, the lifecycle will run this buildpack as the `root` user.
+
+For each root buildpack, the lifecycle will use [snapshotting](https://github.com/GoogleContainerTools/kaniko/blob/master/docs/designdoc.md#snapshotting-snapshotting) to capture changes made during the buildpack's build phase (excluding `/tmp`, `/cnb`, and `/layers`). Alternatively to snapshotting, a platform may store a new stack iamge to cache the changes. All of the captured changes will be included in a single layer produced as output from the buildpack. The `/layers` dir MAY NOT be used to create arbitrary layers.
+
+The buildpack can then exclude directories from the layer by writing to the `launch.toml` using a new `[[excludes]]` table, for example:
+
+```toml
+[[excludes]]
+paths = [ "/var" ]
+cache = true
+```
+
+* `paths` - a list of path globs to exclude from the run image
+* `cache` - if set to `true`, the paths will be stored in the cache
+
+The `[process]` and `[[slices]]` tables can be used as normal.
+
+The following constraints will be enforced:
+* If a root buildpack is included in a `[[build.buildpacks]]` list in an app's `project.toml`, the build will fail.
+* If a user attempts to create a buildpackage including both root buildpacks and non-root buildpacks, the process will fail.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Building an App with Additional Packages
+
+Given a root buildpack with ID `example/git`, the following configuration can be used with a Golang app:
+
+`project.toml`:
+```toml
+[[build.buildpacks]]
+id = "example/git"
+version = "1.0"
+
+[[build.buildpacks]]
+id = "example/golang"
+version = "1.0"
+```
+
+When the following command is run, the app source code will be loaded into the image for access by the root buildpack.
+
+`pack build sclevine/myapp`
+
+All root buildpacks will be sliced out of the list of buildpacks, and run before non-root buildpacks. This creates a version of the current builder with additional layers and an ephemeral run image with additional layers, then does a normal `pack build`.
+
+* The run image layers are persistented as part of the app image, and can be reused on subsequent builds.
```

Comment:
Yes, the root buildpacks must run twice: once on the build image and once on the run image. I'm not sure about reusing the run-image layer. It might not be possible.

Reply:
I think we should cache the previous run image layers locally if they can be reused. They need to be built locally anyways.